source/server/options_impl.h

```diff
@@ -144,6 +145,28 @@ class OptionsImpl : public Server::Options, protected Logger::Loggable<Logger::I
   uint32_t count_;
 };
 
+/**
+ * Interface allowing extensions to add command-line flags to be parsed, by registering an provider
+ * using Registry::RegisterFactory<CommandLineFlagProvider>.
+ */
+class CommandLineFlagProvider {
+public:
+  virtual ~CommandLineFlagProvider() {}
+
+  /**
+   * Method called by OptionsImpl on any registered providers prior to parsing
+   * command line. Provider implementations should call Add() on the given
+   * CmdLine instance, to register additional flags to be parsed.
+   * @param cmdline CmdLine instance that will be used to parse command line.
+   */
+  virtual void AddFlags(TCLAP::CmdLineInterface& cmdline) PURE;
```

Comment:
Two uncoordinated CommandLineFlagProviders may try to add flags with the same name, do we need to handle such errors?